### DISCUSSION TITLE
Playback artifacts on AMD GPU

### DISCUSSION THREAD

**Discussion by owochle** [jellyfin/jellyfin#12745] [Q&A]
Hi,
Since switching to an AMD GPU, i've encountered regular visual artifacts during playback, whatever client I use. The artifacts are blocky ghost images, staying from one frame to the other. Sometime staying for as long as 10 seconds. There are also some jumps in the video, the timeline is still going up but the video is jumping back.

![image](https://github.com/user-attachments/assets/e69014c4-1cab-4460-9ddf-3eccfbbf24e6)
An example of the artifacts

System infos :
```
uname -a: Linux ditto 6.10.11-1.el9.elrepo.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Sep 18 14:54:58 EDT 2024 x86_64 x86_64 x86_64 GNU/Linux
CPU : Intel(R) Core(TM) i3-8100 CPU @ 3.60GHz
GPU : AMD RX 5600XT
```

**Comment by gnattu** [comment]
Can you post ffmpeg log for this? This could be some deficiencies of AMD's H264 encoder and can be workaround by either increate bitrate or use HEVC

**Comment by owochle** [comment]
I'm using VAAPI, HEVC encoding is enabled but for an unknown reason, always transcoded even in source quality.
How do I gather ffmpeg logs in the docker container ?

**Reply by gnattu** [reply]
Dashboard->Logs

**Reply by owochle** [reply]
Here are the logs 
[Log.txt](https://github.com/user-attachments/files/17179106/Log.txt)

**Reply by gnattu** [reply]
Yep, it is using AMD H264 encoder. I'm a bit surprised because 8000k bitrate for 24fps 1080p video should be high enough to not having such artifacts. Probably it is another bug caused by problematic VBR controlling. 

Can you try running this in your docker shell:

```shell
/usr/lib/jellyfin-ffmpeg/ffmpeg -analyzeduration 200M -probesize 1G -init_hw_device drm=dr:/dev/dri/renderD128 -init_hw_device vaapi=va@dr -init_hw_device vulkan=vk@dr -filter_hw_device vk -hwaccel vaapi -hwaccel_output_format vaapi -noautorotate -canvas_size 1920x1080 -i file:"/media/shows/House (2014-2012) [tt0412142]/Season 1/House (2004) - S01E03 - Occam's Razor (1080p BluRay x265 Panda).mkv" -noautoscale -map_metadata -1 -map_chapters -1 -threads 3 -map 0:0 -map 0:1 -map -0:0 -codec:v:0 h264_vaapi -rc_mode CBR -b:v 7865457 -maxrate 7865457 -bufsize 15730914 -profile:v:0 high -sei -a53_cc -force_key_frames:0 "expr:gte(t,n_forced*3)" -filter_complex "[0:2]scale=-1:1080:fast_bilinear,scale,crop,pad=max(1920\,iw):max(1080\,ih):(ow-iw)/2:(oh-ih)/2:black@0,crop=1920:1080,format=bgra,hwupload=derive_device=vulkan,format=vulkan[sub];[0:0]setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709,hwmap=derive_device=vulkan,format=vulkan,libplacebo=upscaler=none:downscaler=none:format=bgra[main];[main][sub]overlay_vulkan=eof_action=pass:repeatlast=0,hwmap=derive_device=vaapi,format=vaapi,scale_vaapi=format=nv12" -start_at_zero -codec:a:0 libfdk_aac -ac 2 -ab 256000 -ar 48000 -af "volume=2" -copyts -avoid_negative_ts disabled -max_muxing_queue_size 2048 test.mp4
```

And get the `test.mp4` to see if it still has such artifacts? If so we may force AMD H264 to use CBR as well.

You can also try a browser that supports HEVC like chrome, and force a transcoding by lowering the bitrate to 6mbps to see if such problem still exits when using AMD HEVC encoder.

**Reply by owochle** [reply]
I don't have access to a computer right now, I'll try as soon as I can. However, does the app supports HEVC, I've been using it too and having the same kind of artifacts on it. I'm on a Google Pixel 8 Pro which should support HEVC.

**Reply by gnattu** [reply]
If the logs says it is using H264 it is using H264. You can check that by playing it on your phone, look at the log and see if it is using `h264_vaapi` or `hevc_vaapi`

**Reply by owochle** [reply]
Yes, the command you told me to test leads to artifacts.

> If the logs says it is using H264 it is using H264. You can check that by playing it on your phone, look at the log and see if it is using `h264_vaapi` or `hevc_vaapi`

Yes it's using h264 on my phone which supposedly supports HEVC.

**Reply by gnattu** [reply]
Interesting, so even CBR cannot fix AMD's H264 vaapi encoder...

What about this, which uses hevc?

```
/usr/lib/jellyfin-ffmpeg/ffmpeg -analyzeduration 200M -probesize 1G -init_hw_device drm=dr:/dev/dri/renderD128 -init_hw_device vaapi=va@dr -init_hw_device vulkan=vk@dr -filter_hw_device vk -hwaccel vaapi -hwaccel_output_format vaapi -noautorotate -canvas_size 1920x1080 -i file:"/media/shows/House (2014-2012) [tt0412142]/Season 1/House (2004) - S01E03 - Occam's Razor (1080p BluRay x265 Panda).mkv" -noautoscale -map_metadata -1 -map_chapters -1 -threads 3 -map 0:0 -map 0:1 -map -0:0 -codec:v:0 hevc_vaapi -rc_mode CBR -b:v 7865457 -maxrate 7865457 -bufsize 15730914 -profile:v:0 main -force_key_frames:0 "expr:gte(t,n_forced*3)" -filter_complex "[0:2]scale=-1:1080:fast_bilinear,scale,crop,pad=max(1920\,iw):max(1080\,ih):(ow-iw)/2:(oh-ih)/2:black@0,crop=1920:1080,format=bgra,hwupload=derive_device=vulkan,format=vulkan[sub];[0:0]setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709,hwmap=derive_device=vulkan,format=vulkan,libplacebo=upscaler=none:downscaler=none:format=bgra[main];[main][sub]overlay_vulkan=eof_action=pass:repeatlast=0,hwmap=derive_device=vaapi,format=vaapi,scale_vaapi=format=nv12" -start_at_zero -codec:a:0 libfdk_aac -ac 2 -ab 256000 -ar 48000 -af "volume=2" -copyts -avoid_negative_ts disabled -max_muxing_queue_size 2048 test.mp4
```

> Yes it's using h264 on my phone which supposedly supports HEVC.

You may need to change your android client's player to exoplayer instead of the web player to see hevc.

**Reply by owochle** [reply]
> You may need to change your android client's player to exoplayer instead of the web player to see hevc.

Thanks for the tip. Why isn't it enabled by default since it has more features (like HDR playback etc....) ?

And about the command, I still have artifacts at the same places. Just to be sure, I also checked the source file from which ffmpeg is transcoding and it has no issues so I'm clueless.

**Reply by gnattu** [reply]
> it has no issues so I'm clueless.

I don't know but it really looks like AMD issue or at least their driver issue to me. This does not happen if you just use your integrated intel iGPU for transcoding right?

**Reply by owochle** [reply]
Sadly, my bios doesn't let me use both at the same time so I can't test this (I don't have physical access to the machine for at least 2-3 weeks).

**Reply by owochle** [reply]
So, this issue is unsolvable then ? I don't know what happened, but it got worse just by moving the graphics card, so I think it may be the pcie extender I'm using. Could it be it/the card not plugged correctly ?
